### PR TITLE
WIP: Add an embedded QR code scanner

### PR DESCRIPTION
**Pull request by katcom**
I have improve this app by adding a embedded QR code scanner, so that users can easily add friends with QR code without a extra application to be installed.  
Also, some code in MainActivity and AddFriendActivity were changed, for the purpose of adding a dialog that reminds users the QR code has been read successfully, and making the app to add friend directly after the scan of QR code without pressing the add friend button.

### CODE REVIEW

**Comment by katcom** [comment]
Any advice?

File: android-refimpl-app/app/build.gradle [hunk 1]
```diff
@@ -19,7 +19,8 @@
 buildscript {
     repositories {
         maven {
-            url 'https://maven.google.com/'
+            //url 'https://maven.google.com/'
+            url 'https://dl.google.com/dl/android/maven2/'
```
Comment (zoff99) [inline android-refimpl-app/app/build.gradle:23]: what is this change for?

File: android-refimpl-app/app/src/main/java/com/zoffcc/applications/trifa/AddFriendActivity.java [hunk 2]
```diff
@@ -104,48 +109,46 @@ public void read_qr_code(View v)
     {
         try
         {
-            Intent intent = new Intent("com.google.zxing.client.android.SCAN");
-            intent.putExtra("SCAN_MODE", "QR_CODE_MODE"); // "PRODUCT_MODE for bar codes
+            // Disuse the API from zxing client
```
Comment (zoff99) [inline android-refimpl-app/app/src/main/java/com/zoffcc/applications/trifa/AddFriendActivity.java:112]: remove unused code lines, not just comment them out

File: android-refimpl-app/app/src/main/java/com/zoffcc/applications/trifa/MainActivity.java [hunk 3]
```diff
@@ -1008,12 +1011,21 @@ else if (position == 8)
...
                                 }
                                 else
                                 {
                                     // just exit
                                     tox_service_fg.stop_me(true);
                                 }
+
+                                if(Build.VERSION.SDK_INT>=16 && Build.VERSION.SDK_INT<21){
+                                    //getActivity.finishAffinity();
+                                    ActivityCompat.finishAffinity(MainActivity.this);
```
Comment (zoff99) [inline android-refimpl-app/app/src/main/java/com/zoffcc/applications/trifa/MainActivity.java:1024]: why was this added, what does it do?

File: android-refimpl-app/app/src/main/java/com/zoffcc/applications/trifa/MainActivity.java [hunk 4]
```diff
@@ -1008,12 +1011,21 @@ else if (position == 8)
...
+                                } else if(Build.VERSION.SDK_INT>=21){
+                                    //getActivity().finishAndRemoveTask();
+                                    ActivityCompat.finishAffinity(MainActivity.this);
```
Comment (zoff99) [inline android-refimpl-app/app/src/main/java/com/zoffcc/applications/trifa/MainActivity.java:1027]: why was this added, what does it do?

File: android-refimpl-app/app/src/main/res/layout/dialog_friend_request_success.xml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        android:text="@string/friend_request_sent"
+        android:visibility="gone"/>
+
+    <ImageView
+        android:id="@+id/dialog_friend_request_success_ok_button"
+        android:layout_width="50dp"
+        android:layout_height="50dp"
+        android:layout_margin="20dp"
+        android:layout_gravity="center_horizontal"
+        app:srcCompat="@drawable/ic_ok"/>
```
Comment (zoff99) [inline android-refimpl-app/app/src/main/res/layout/dialog_friend_request_success.xml:27]: please dont add new graphics, we already have those. look at the Video call screen for reference

File: android-refimpl-app/app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -195,6 +195,10 @@
     QR code button to scan friend\'s code with a camera.</string>
     <string name="add_friend_activity_add">Add Friend</string>
     <string name="cancel">Cancel</string>
+
+    <string name="friend_request_sent"> Friend request has been sent.</string>
```
Comment (zoff99) [inline android-refimpl-app/app/src/main/res/values/strings.xml:199]: is the " "  at the start needed?

**Comment by zoff99** [comment]
@katcom please see comments

**Comment by katcom** [comment]
It looks strange that the test hasn't completed in two days.

**Comment by zoff99** [comment]
DEP is not working anymore. ignore it

**Comment by JeanPaulLucien** [comment]
I like this feature. Add it in night build and compile APK for testing.

**Comment by katcom** [comment]
Hey, Could you help improve this pull request, since I  am not very sophisticated in testing?

**Comment by katcom** [comment]
By the way, the code scanner is fully functional.

**Review by zoff99 — CHANGES_REQUESTED** [review]

File: android-refimpl-app/app/build.gradle (hunk 1 above)
Comment (katcom) [reply android-refimpl-app/app/build.gradle:23]: Maybe it can be discard. Actually https://maven.google.com/ will be redirected to https://dl.google.com/dl/android/maven2/. But former is not available in China, but the latter is.

File: android-refimpl-app/app/src/main/res/values/strings.xml (hunk 6 above)
Comment (katcom) [reply android-refimpl-app/app/src/main/res/values/strings.xml:199]: It just looks nicer in the app. It is actually not used, the visibility of text view using it is gone since I tried a couple of layouts, and found that a green yes icon is most intuitive to users. I simply forgot to delete the unused text view and this string. Sorry for the confusstion.

File: android-refimpl-app/app/build.gradle (hunk 1 above)
Comment (zoff99) [reply android-refimpl-app/app/build.gradle:23]: it needs to work on the f-droid build server

Comment (katcom) [reply android-refimpl-app/app/build.gradle:23]: Ok, just discard this particular change. This alternation simply was for the chinese great fire wall that blocks maven.google.com.

**Comment by quantumpacket** [comment]
Any update on this?

**Comment by zoff99** [comment]
the requested changes were not made.
also we not not want to maintain another piece of code.
